src/js/components/LeftPane.tsx

```diff
@@ -284,21 +295,138 @@ function HistorySection({isOpen, style, resizeProps, toggleProps}) {
+  const queryResults = []
+  new TreeModel({childrenPropertyName: "items"})
+    .parse(queriesRoot)
+    .walk((n) => {
+      if (!n.model.tags) return true
+      if (includes(n.model.tags, tag)) queryResults.push(n.model)
+
+      return true
+    })
+
+  return queryResults
+}
+
 function QueriesSection({isOpen, style, resizeProps, toggleProps}) {
-  const root = useSelector(Queries.getRaw)
+  const dispatch = useDispatch()
+  const [contextArgs, setContextArgs] = useState(null)
+  const [selectedTag, setSelectedTag] = useState("All")
+  const currentSpace = useSelector(Current.getSpace)
+  const queriesRoot = useSelector(Queries.getRaw)
+  const [queries, setQueries] = useState(queriesRoot)
+  const tags = useSelector(Queries.getTags)
+  const hasMultiSelected = contextArgs && contextArgs.selections.length > 1
+
+  useEffect(() => {
+    setQueries(queriesRoot)
+  }, [queriesRoot])
+
+  const template: MenuItemConstructorOptions[] = [
+    {
+      label: "Run Query",
+      enabled: !hasMultiSelected,
+      click: () => {
+        if (!currentSpace)
+          return dispatch(Notice.set(new Error("No space selected")))
+
+        const {
+          item: {value}
+        } = contextArgs
+
+        dispatch(SearchBar.clearSearchBar())
+        dispatch(SearchBar.changeSearchBarInput(value))
+        dispatch(Search.setSpanArgs(brim.space(currentSpace).everythingSpan()))
+        dispatch(submitSearch())
+      }
+    },
+    {
+      label: "Copy Query",
+      enabled: !hasMultiSelected,
+      click: () => {
+        const {
+          item: {value}
+        } = contextArgs
+        lib.doc.copyToClipboard(value)
+        popNotice("Query copied to clipboard")
+      }
+    },
+    {type: "separator"},
+    {
+      label: "Edit",
+      enabled: !hasMultiSelected,
+      click: () => {
+        const {item} = contextArgs
+        // only edit queries
+        if ("items" in item) return
+        dispatch(Modal.show("edit-query", {query: item}))
+      }
+    },
+    {type: "separator"},
+    {
+      label: hasMultiSelected ? "Delete Selected" : "Delete",
+      click: () => {
+        return remote.dialog
+          .showMessageBox({
+            type: "warning",
+            title: "Confirm Delete Query Window",
+            message: `Are you sure you want to delete the ${(contextArgs.selections &&
+              contextArgs.selections.length) ||
+              ""} selected quer${hasMultiSelected ? "ies" : "y"}?`,
+            buttons: ["OK", "Cancel"]
+          })
+          .then(({response}) => {
+            if (response === 0) {
+              const {selections, item} = contextArgs
+              if (hasMultiSelected) dispatch(Queries.removeItems(selections))
+              else dispatch(Queries.removeItems([item]))
+            }
+          })
+      }
+    }
+  ]
 
-  function onItemClick(e, item) {
-    console.log("clicked", item)
+  const menu = usePopupMenu(template)
```

Comment:
I know I'm the one who wrote this, but I think this is pattern for context menus is making our app slow. Anytime data needed only in the context menu gets updated, this whole component has to re-render.

A better pattern is to wait until the user has right clicked, then dispatch a thunk. Then in the thunk we gather all the state needed for the context menu and show it.